### PR TITLE
Downloading/extracting of Maven binaries is thread-safe

### PR DESCRIPTION
**Pull request by MatousJobanek**

### CODE REVIEW

File: maven/impl-maven-embedded/pom.xml [hunk 1]
```diff
@@ -77,6 +77,12 @@
             <version>${version.awaitility}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.github.stefanbirkner</groupId>
+            <artifactId>system-rules</artifactId>
+            <version>1.16.0</version>
```
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/pom.xml]: Detail - make it a property? :)

File: maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
 public abstract class DistributionStageImpl<NEXT_STEP extends BuildStage<DAEMON_TRIGGER_TYPE>, DAEMON_TRIGGER_TYPE extends DaemonBuildTrigger>
     implements DistributionStage<NEXT_STEP, DAEMON_TRIGGER_TYPE> {
 
-    private String maven3BaseUrl =
+    private static String maven3BaseUrl =
```
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java]: UPPER_CASE if const? Could it be `final` though?

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java [hunk 3]
```diff
@@ -56,6 +64,53 @@ public void testDownloadFromGivenDestinationWithoutUsingCacheAndExtract() throws
...
+    public void testDownloadInMultipleThreads() throws IOException, InterruptedException {
+        // cleanup
+        FileUtils.deleteDirectory(targetMavenDir);
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        // multiple download
+        createThreadWithDownload(latch).start();
+        createThreadWithDownload(latch).start();
+        latch.countDown();
+        downloadSWR300Alpha1();
```
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: Shouldn't we do the actual download in the threads? Also `countDown` doesn't look to me as it's in the right place.

Comment (MatousJobanek) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: It is lazy (in the sense that I'm lazy) solution for waiting till all the downloads are finished. But you are right - I'll do it properly

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java [hunk 4]
```diff
@@ -56,6 +65,60 @@ public void testDownloadFromGivenDestinationWithoutUsingCacheAndExtract() throws
         verifyExtraction(2, "resolver-3.0.0-alpha-1", "resolver-3.0.0-alpha-3");
     }
 
+    @Test
+    public void testDownloadInMultipleThreads() throws IOException, InterruptedException {
+        // cleanup
+        FileUtils.deleteDirectory(targetMavenDir);
+
+        // multiple download
+        CountDownLatch latch = new CountDownLatch(1);
```
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: How about making it `new CountDownLatch(2)` and then in the thread itself:

```java
private Thread createThreadWithDownload(final CountDownLatch latch) {
         return new Thread(new Runnable() {
             @Override
             public void run() {
                 try {
                     latch.await();
                     downloadSWR300Alpha1();
                     latch.countDown();
                 } catch (InterruptedException e) {
                     e.printStackTrace();
                 }
             }
         });
     }
```

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java [hunk 5]
```diff
@@ -56,6 +65,60 @@ public void testDownloadFromGivenDestinationWithoutUsingCacheAndExtract() throws
...
+                    latch.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+                downloadSWR300Alpha1();
+            }
+        });
+    }
+
+    private void downloadSWR300Alpha1() {
```
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: I'm not sure if revealing version in the method name is worth to have - if we gonna have new version chances are method name will stay the same and thus become stale :)

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java [hunk 6]
```diff
@@ -56,6 +65,60 @@ public void testDownloadFromGivenDestinationWithoutUsingCacheAndExtract() throws
...
+        try {
+            EmbeddedMaven
+                    .forProject(pathToJarSamplePom)
+                    .useDistribution(new URL("http://github.com/shrinkwrap/resolver/archive/3.0.0-alpha-1.zip"), false);
```
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: I was also wondering - doesn't it make this test vulnerable for network issues? Could we somehow simulate download process instead so we have it all locally?

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java (hunk 4 above)
Comment (MatousJobanek) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: well, this wouldn't do the job - I need to start both threads at once - at the very same time.
In addition - your code would freeze - you are waiting for a latch that will be counted down in a future by the same code...

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java (hunk 6 above)
Comment (MatousJobanek) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: sure, but not part of this PR...

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java (hunk 5 above)
Comment (MatousJobanek) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: The reason why I used the version there is that there are downloads of different version in this class...

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java (hunk 6 above)
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: Sure, didn't say as part of it, but let's create an issue for that.

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java (hunk 5 above)
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: Maybe make one method with parameter instead?

Comment (MatousJobanek) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: yeah, right... that could work

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java (hunk 4 above)
Comment (bartoszmajsak) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: Correct, I was too fast, we would need two latches for that.

Comment (MatousJobanek) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java]: yeah, two latches or wait until both threads are stopped. https://github.com/shrinkwrap/resolver/pull/120/files/88f2b2f8b038a8facec04866da2b4a574367d2bd#diff-9df706518aa71a393efa82332a01b03aR83
two latches would be probably less of code...

**Comment by MatousJobanek** [comment]
fixed. 
and with two latches it is much better. And also much easier to add another download...
Thx for proper review and good discussion - I don't use threading often so it's not one of my strengths.
